### PR TITLE
Fix for ScrollContainer does not save position on page reload

### PR DESCRIPTION
**Pull request by alexhisen**
Make element scroll position be saved immediately on scroll to match behavior of window and to allow state to be preserved on page reload.

Add a getSavedScrollTarget() method for tests and also to enable shouldUpdateScroll function to know if there is a state that will be restored.

### CODE REVIEW

**Comment by alexhisen** [comment]
I've updated tests, so the patch should have 100% coverage now. Don't know why FF is failing on Travis. FF 56 on Windows passes all tests just fine.

**Comment by taion** [comment]
The tests are super flaky especially on FF, sorry. A bit busy today but will look soon.

File: src/index.js [hunk 1]
```diff
@@ -57,14 +57,30 @@ export default class ScrollBehavior {
     });
   }
 
+  getSavedScrollTarget(key) {
```
Comment (taion) [inline src/index.js]: Can you make this a separate PR?

File: src/index.js [hunk 2]
```diff
@@ -57,14 +57,30 @@ export default class ScrollBehavior {
...
   registerElement(key, element, shouldUpdateScroll, context) {
     invariant(
       !this._scrollElements[key],
       'ScrollBehavior: There is already an element registered for `%s`.',
       key,
     );
 
-    this._scrollElements[key] = { element, shouldUpdateScroll };
+    const onElementScroll = () => {
+      this._saveElementPosition(key);
```
Comment (taion) [inline src/index.js]: I think we should rAF-throttle these. They're still writing to local storage. Note in that case you'd have to apply the same logic then to cancel pending timers to save element positions.

File: src/index.js (hunk 1 above)
Comment (alexhisen) [reply src/index.js]: I need it in the same PR to be able to test the new methods.

File: src/index.js (hunk 2 above)
Comment (alexhisen) [reply src/index.js]: see updated code

**Comment by alexhisen** [comment]
Can you please merge this PR ASAP.

**Comment by taion** [comment]
I will. I'm a little busy right now but I'll get to this in the next day or two.

**Comment by alexhisen** [comment]
Any progress? :-)

**Comment by taion** [comment]
Buried under deadlines this week. Soon!

**Comment by alexhisen** [comment]
How is it going?

**Comment by alexhisen** [comment]
Yeah, I guess it's fine. Should I submit a separate PR for getSavedScrollTarget()?

**Comment by taion** [comment]
Sure, but it should work a bit differently. Ideally we wrap up separately a `getWindowScrollTarget(prevContext, context)` and a `getElementScrollTarget(key, prevContext, context)`.

**Comment by taion** [comment]
Released in v0.9.4.

**Comment by alexhisen** [comment]
I agree that a getSavedWindowScrollTarget and getSavedElementScrollTarget is probably better but I also use this method outside of shouldUpdateScroll so I don't have contexts available to me. I use it at component mount to know if scroll is going to be restored for me and if not, I need to scroll to a certain position that's not 0,0. In shouldUpdateScroll I suppress the default 0,0 by returning false if there is not a saved one. These happen in different components, so shouldUpdateScroll doesn't know where to scroll to, only that it shouldn't restore to default.

The sequence of which fires first is also unpredictable, hence the use of the method to know whether scroll restore will happen or has already happened. I access this method from the undocumented scrollBehavior context.
